stumpy/core.py

```diff
@@ -2561,6 +2575,163 @@ def _select_P_ABBA_value(P_ABBA, k, custom_func=None):
     return MPdist
 
 
+@njit
+def _merge_topk_PI(PA, PB, IA, IB):
+    """
+    Merge two top-k matrix profiles `PA` and `PB`, and update `PA` (in place) while
+    always prioritizing the values of `PA` over the values of `PB` in case of ties.
+    (i.e., values from `PB` are always inserted to the right of values from `PA`).
+    Also, update `IA` accordingly.
+
+    Unlike `_merge_topk_ρI`, where `top-k` largest values are kept, this function
+    keeps `top-k` smallest values.
+
+    Parameters
+    ----------
+    PA : numpy.ndarray
+        A (top-k) matrix profile where values in each row are sorted in ascending
+        order. `PA` must be 2-dimensional.
+
+    PB : numpy.ndarray
+        A (top-k) matrix profile where values in each row are sorted in ascending
+        order. `PB` must have the same shape as `PA`.
+
+    IA : numpy.ndarray
+        A (top-k) matrix profile indices corresponding to `PA`
+
+    IB : numpy.ndarray
+        A (top-k) matrix profile indices corresponding to `PB`
+
+    Returns
+    -------
+    None
+    """
+    k = PA.shape[1]
+    tmp_P = np.empty(k, dtype=np.float64)
+    tmp_I = np.empty(k, dtype=np.int64)
+    for i in range(PA.shape[0]):
+        overlap = np.intersect1d(IA[i], IB[i])
+        aj, bj = 0, 0
+        idx = 0
+        for _ in range(2 * k):  # 2 * k to traverse both A and B
+            if idx >= k:
+                break
+
+            if bj < k and PB[i, bj] < PA[i, aj]:
+                if IB[i, bj] not in overlap:
+                    tmp_P[idx] = PB[i, bj]
+                    tmp_I[idx] = IB[i, bj]
+                    idx += 1
+                bj += 1
+
+            else:
+                tmp_P[idx] = PA[i, aj]
+                tmp_I[idx] = IA[i, aj]
+                idx += 1
+                aj += 1
+
+        PA[i] = tmp_P
+        IA[i] = tmp_I
+
+
+@njit
+def _merge_topk_ρI(ρA, ρB, IA, IB):
+    """
+    Merge two top-k pearson profiles `ρA` and `ρB`, and update `ρA` (in place) while
+    always prioritizing the values of `ρA` over the values of `ρB` in case of ties.
+    (i.e., values from `ρB` are always inserted to the left of values from `ρA`).
+    Also, update `IA` accordingly.
```

Comment:
This function is used in `stumpy.stump()` to merges the results that are obtained from different threads). And, we do not have the overlapping-values issue there. In other words, the inputs `IA[i]` and `IB[i]` do not have any common elements. However, for the sake of consistency, we may want to revise this function to avoid duplicates. (so that if we use it later somewhere else, we would be safe). 

What do you think?

Reply:
I am fine with that. It would be confusing to not see the overlap check in this function so let's keep both `merge` functions as similar/consistent as possible especially if there is very little performance cost.

Reply:
> It would be confusing to not see the overlap check in this function so let's keep both `merge` functions as similar/consistent as possible especially if there is very little performance cost.

Yes...exactly. I will ran the performance test as well.

Reply:
@seanlaw 
So, I added the overlap check to `core._merge_topk_ρI`, and I ran the performance test. please see table below.
```
# input
seed = 0
np.random.seed(seed)

T = np.random.rand(100_000)
m = 50
```


| stumpy.stump |without overlap check| with overlap check|
|:----|:----|:----|
|1 (main)|17.9|17.9|
|1|14.1|16|
|5|36.5|41|
|10|64|68|
|15|84.9|89|
|20|101.1|112|
|100|215.5|231|

What do you think?

Reply:
Can you please show the code here? What is the difference between `1 (main)` and `1`? Why are their timings different?

Reply:
The "main" in `1 (main)` is regarding the  branch `main` that only supports `k=1`. I just  provided the number in the table just in case we want to compare the performance of top-1 of this branch with branch `main`.

```
# performance code
# stump
stumpy.stump(np.random.rand(10), 3, k=1) 

k_lst = [1, 5, 10, 15, 20] # k_lst = [100]

ComputeTime =  []
for k in k_lst:
    print(k, end='-->')
    tic = time.time()
    stumpy.stump(T, m, k=k)
    toc = time.time()
    ComputeTime.append(toc - tic)
```

And, the code for `_merge_topk_ρI` with check for overlap:
```
# in stumpy/core.py
def _merge_topk_ρI(ρA, ρB, IA, IB):
    k = ρA.shape[1]
    tmp_ρ = np.empty(k, dtype=np.float64)
    tmp_I = np.empty(k, dtype=np.int64)
    last_idx = k - 1
    for i in range(len(ρA)):
        overlap = np.intersect1d(IA[i], IB[i])
        aj, bj = last_idx, last_idx
        idx = last_idx
        for _ in range(2 * k):
            if idx < 0:
                break
            if ρB[i, bj] > ρA[i, aj]:
                if bj >= 0 and IB[i, bj] not in overlap:
                    tmp_ρ[idx] = ρB[i, bj]
                    tmp_I[idx] = IB[i, bj]
                    idx -= 1
                bj -= 1
            else:
                tmp_ρ[idx] = ρA[i, aj]
                tmp_I[idx] = IA[i, aj]
                idx -= 1
                aj -= 1

        ρA[i] = tmp_ρ
        IA[i] = tmp_I
```

Reply:
> Can all of the additional time be attributed to checking the overlap in core._merge_topk_ρI or is there a possibility that it is coming from elsewhere?

I am trying to figuring this out :) I THINK we might actually have the overlapping thing in the stumpy.stump as well. I am doing some investigation. Will update you.

Reply:
Why are you doing `for _ in range(2 * k):`?

Reply:
I wanted to avoid while, and thus 2*k is the maximum number of iterations we need to fill in the top-k `tmp_P` and `tmp_I` arrays. I wanted to take care of any rare case that might happen. 

For instance, let's take a look at the following scenario. This is for `_merge_topk_PI` where we traverse from left to right. But, the concept/reason is the same.

```
# top-2 of seq `i` from a thread
PA[i] =[1+e-8,  1 + e-7]
IA[i]  = [0, 2]

# top-2 of seq `i` from another thread
PB[i] =[1+e-10, 1 + e-9]
IB[i] = [0, 1]

# overlap = [0]
```

So, in the example above, `PB[i,0] < PA[i, 0]` but `IB[i,0] is `in overlap`. So, we skip it, and we are not going to use PA[i, 0]. So, we go to the next iteration while we haven't updated `tmp_P` and `tmp_I` yet. So, if we go with `k` iterations, we may reach the end of iterations while `tmp_P`/`tmp_I` are not completely filled. 

What do you think?

Reply:
The following example is another case, and I think it is a better example:

```
# top-3 of seq `i` from a thread
PA[i] =[0, 2,   3]
IA[i]  = [0, 10, 20]

# top-3 of seq `i` from another thread
PB[i] =[e-9, e-8, 1.5]
IB[i] = [3, 0, 1]

# overlap = [0]
```

We start with `aj=0` and `bj=0` in the `j=0`-th iteration. In the following, I am demonstrating the updating process: 

* We first choose `PA[i, aj=0]` / `IA[i, aj=0]` and we fill `tmp_P[0]` and `tmp_I[0]`. We update `aj`/`bj`: `aj=1` and `bj=0`. 
* Now, we choose `PB[i, bj=0]` / `IB[i, bj=0]` and we fill `tmp_P[1]` and `tmp_I[1]`. We update `aj`/`bj`: `aj=1` and `bj=1`.
* Now, we choose `PB[i, bj=1]` / `IB[i, bj=1]`; HOWEVER, `IB[i, 1]` is an overlapping value. We skip it. We update `aj`/`bj`: `aj=1` and `bj=2`.
* Now, we choose `PB[i, bj=2]` / `IB[i, bj=2]` and we fill `tmp_P[2]` and `tmp_I[2]`. 
* Done!

We had four iterations for `k=3`.

Reply:
@seanlaw 
**UPDATE**

> Can all of the additional time be attributed to checking the overlap in `core._merge_topk_ρI` or is there a possibility that it is coming from elsewhere?

So, I finished my investigation with the following input for `stumpy.stump`.
```
seed = 0
np.random.seed(seed)
n = 100_000

T = np.random.rand(n)
m = 50
k  = [1, 5, 10, 15, 20]

# l = n - m + 1
``` 

**NOTES:**
* I checked if there is any overlap between IA[i] and IB[i] throughout the merging process in `stumpy.stump`. Recall that to merge the results of two threads, we need to merge two matrix profiles (as well as matrix profile indices) of shape `(l, k)`. I checked the overlap between IA[i] and IB[i] and I confirmed that there is no overlap. So, `overlap = np.intersect1d(IA[i], IB[i])` is empty in ALL iterations, and thus the check `if IB[i, bj] not in overlap:` should be fast. 

**important note**:  Although there is no overlapping value (between `IA[i]` and `IB[i]`) in this example, that may not be always the case. It is possible that `IA[i]` and `IB[i]` have the element `-1` in common.  Reminder: the `IA[i, index] = -1` corresponds to `ρA[i, index] = np.NINF`. So, if we have an overlapping value, that overlapping value can only be `-1`! So, the length of array `overlap` is one at the most!

* I replaced `overlap = np.intersect1d(IA[i], IB[i])` with `np.empty(shape=(0, ), dtype=np.int64)`, and ran the performance test again. I got (almost) the same computing time as when there was no check for `overlap`. So, the increase in computing time is caused by `np.intersect1d(IA[i], IB[i])`. 

* It seems that the function `np.intersect1d()` has an argument `assume_unique`, which, if set to `True`, can speed up the computing process (I took a look at its [code](https://github.com/numpy/numpy/blob/v1.23.0/numpy/lib/arraysetops.py#L373-L469). When `assume_unique` is set to True, it does not execute `numpy.unique()`. There are also many other unnecessary things that we do not need here) Unfortunately, the argument `assume_unique` is not supported in numba. (maybe I couldn't figure out how to do it!)  **So, I wrote a code to get the overlapping values between two arrays. This is adapted from the code of `numpy.intersect1d()`.**

```
@njit
def _intersect(ar1, ar2):
    """
    finding the non-negative overlapping values between two 1D arrays `ar1` and `ar2`, where the non-negative 
    values in each array is unique.
    """
    aux = np.concatenate((ar1, ar2))
    aux = aux[aux >= 0] # we do not need to find overlapping value `-1`
    aux.sort()

    mask = aux[1:] == aux[:-1]
    overlap = aux[:-1][mask]

    return overlap
```

I am comparing the performances below (please do not compare the following numbers with previous ones as the pc that I am using right now is different)

|k |no overlap check|njit function to get overlap |np.intersect|
|:----|:----|:----|:----|
|1|20.23|22.64|24.27|
|5|36.3|37.4|38.78|
|10|64.6|65.88|67.26|
|50|153.86|157.97|160.05|

```
# performance code
seed  = 0
np.random.seed(seed)
T = np.random.rand(100000)
m = 50

stumpy.stump(np.random.rand(10), m=3, k=1) 

ComputeTime =  []
for _ in range(5):
    k = 1  # 5, 10, 50
    tic = time.time()
    stumpy.stump(T, m, k=k)
    toc = time.time()
    ComputeTime.append(toc - tic)

t = np.mean(ComputeTime[1: ])
```

Reply:
May I propose something slightly faster (in my test) and, IMHO, easier to interpret:

```
@njit
def _intersect(arr1, arr2):
    return np.array(
        list(
            set(arr1).intersection(set(arr2))  # Basic set comparison
        ),
        dtype=np.int64
    )
```

>     aux = aux[aux >= 0] # we do not need to find overlapping value `-1`

When you create a generic, reusable function such as `_intersect`, you should really try to avoid embedding weird conditional side effects like this where you've hidden the fact that the `overlap` cannot include negative values. This is bad practice and isn't obvious unless the end user dives deep into the code. Essentially, you've embedded a nasty assumption that only makes sense in this one use case and therefore this function cannot be reused generically elsewhere since it doesn't return the true `overlap` as most people would expect (i.e., it has a conditional that it only returns the overlap for positive values).

Additionally, can you explain why we need to eliminate `-1` from the overlap? Why isn't it fine to keep `-1` in the `overlap` array if it truly exists in both input arrays?

Reply:
> May I propose something slightly faster (in my test) and, IMHO, easier to interpret:
> 
> ```
> @njit
> def _intersect(arr1, arr2):
>     return np.array(
>         list(
>             set(arr1).intersection(set(arr2))  # Basic set comparison
>         ),
>         dtype=np.int64
>     )
> ```

This is obviously more readable!!! I thought numpy should be faster than built-in function set. I will test it. Thanks for the suggestion! 



> > ```
> > aux = aux[aux >= 0] # we do not need to find overlapping value `-1`
> > ```
> 
> 
> When you create a generic, reusable function such as `_intersect`, you should really try to avoid embedding weird conditional side effects like this where you've hidden the fact that the `overlap` cannot include negative values. This is bad practice and isn't obvious unless the end user dives deep into the code. Essentially, you've embedded a nasty assumption that only makes sense in this one use case and therefore this function cannot be reused generically elsewhere since it doesn't return the true `overlap` as most people would expect (i.e., it has a conditional that it only returns the overlap for positive values).

I totally agree. I should have avoided using such assumption in the function. As you said, I created a TOO specific function that may not be useful and reusable in other cases. 

How about this:
```
@njit
def _intersect_unique(ar1, ar2):
    """
    Finds the overlapping values between two 1D arrays `ar1` and `ar2`, where each array consists of unique values.
    This is equivalent to `np.intersect1d(ar1, ar2, assume_unique=True)`
    """
    aux = np.concatenate((ar1, ar2))
    aux.sort()

    mask = aux[1:] == aux[:-1]
    overlap = aux[:-1][mask]

    return overlap
```

And, then, we do:
```
mask_A = IA[i] >= 0
mask_B = IB[i] >= 0
overlap = _intersect_unique(IA[i, mask_A], IB[i, mask_B])
```

So, here, we assume the reader know that IA[i] and IB[i] can have some elements with the value of `-1` and/or some unique non-negative values. ("some" means zero, or one, or more than one). 

Although this is better than what I suggested before, it still requires user know about IA and IB. Also, `mask_A = IA[i] >= 0` might sound strange!

> Additionally, can you explain why we need to eliminate `-1` from the overlap? Why isn't it fine to keep `-1` in the `overlap` array if it truly exists in both input arrays?

So, it is totally fine to keep `-1` in the `overlap`. However, it does not affect the final result in my implementation. Recall that I only checked overlap when a value of B is chosen over A (see the code snippet below) 

```
# in function `_merge_topk_ρI(ρA, ρB, IA, IB)`
     
     for _ in range(2 * k):
            if idx < 0:
                break
            if bj >= 0 and ρB[i, bj] > ρA[i, aj]: # choosing B over A 
                if IB[i, bj] not in overlap:
```

So, when the condition in `if bj >= 0 and ρB[i, bj] > ρA[i, aj]:` becomes True, it means  ρB[i, bj] is definitely not `np.NINF`, and thus `ρB[i, bj]` cannot be `-1`. Note that `-1` in `IB` corresponds to `np.NINF` in ρB. Therefore, we are  not going to see `if -1 not in overlap` in the merging process.

As an aside, according to the algo of  `numpy.intersection(x, y)`, we need to perform `np.unique()` first to extract unique values from each array. However, we can skip it if the values of each array are unique already. So, I wanted to avoid executing `np.unique()`. So, I just excluded `-1` from the arrays. Now, I am sure that each array consists of unique values.



> So, after we find the `mask` (i.e., a boolean array that gets the indices in `IB[i]` that are in `IA[i]`), then we sub-select the specific columns from `B` and process only those columns. This way, we keep the original algorithm (of choosing the smallest `k` from two arrays) clean and easily understandable.

mask is definitely better particularly if we have a large-size array `overlap`. However, apparently `np.in1d` is not supported by numba. So, I took a  look at its source code, and it seems it is similar to `np.intersect1d` to some extent. It executes `np.unique()` unless we set `assume_unique` to True. So, we probably(?) are going to see the same performance. I will test it.

Reply:
> How about this:

I don't know. It feels strange to use `concatenate` and `sort`.

> However, apparently np.in1d is not supported by numba.

I don't know. This worked for me:

```
njit(parallel=True)
def _isin(arr1, arr2):
    return np.in1d(arr1, arr2)
```

At the end of the day, we need to properly balance reducing complexity vs increasing performance. In case it matters, the performance numbers aren't that bad. We may be overthinking it. In that case, we should opt for the simplest solution

Reply:
> I don't know. This worked for me:

oh! maybe I made a mistake. I will try again.

> we need to properly balance reducing complexity vs increasing performance. In case it matters, the performance numbers aren't that bad. We may be overthinking it. In that case, we should opt for the simplest solution

> we need to properly balance reducing complexity vs increasing performance. In case it matters, the performance numbers aren't that bad. We may be overthinking it. In that case, we should opt for the simplest solution

I will try both `built-in set` and `np.in1d` (I hope it works for me as well) and then I will update files and push the new changes. I will also let you know the performance so we can decide whether we should change it again or not. I am going to resolve this conversation for now. 

Sounds good?

Reply:
> `np.in1d` (I hope it works for me as well)

I wrote the function in a script.py file, and it works! (But, it does not work when I write and execute it in Jupiter!)

Reply:
Let's ignore `in1d` for now as it doesn't sound like it is backwards compatible with different versions of `numba` and might cause older versions to fail/break.